src/python/platforms/android/emulator.py

```diff
@@ -0,0 +1,110 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helper functions for running commands on emulated Android devices."""
+
+import os
+import re
+import subprocess
+import tempfile
+import time
+
+from google_cloud_utils import storage
+from metrics import logs
+from platforms.android import adb
+from system import archive
+from system import environment
+from system import new_process
+from system import shell
+
+_WAIT_SECONDS = 30
+
+# Output pattern to parse stdout for serial number
+DEVICE_SERIAL_RE = re.compile(r'DEVICE_SERIAL: (.+)')
+
+
+class EmulatorError(Exception):
+  """Error for errors handling the Android emulator."""
+
+
+class EmulatorProcess(object):
+  """A EmulatorProcess encapsulates the creation, running, and destruction
+  of Android emulator processes."""
+
+  def __init__(self):
+    self.process_runner = None
+    self.process = None
+    self.logfile = None
+
+    log_path = os.path.join(tempfile.gettempdir(), 'android-emulator.log')
+    self.logfile = open(log_path, 'wb')
+
+  def create(self):
+    """Configures a emulator process which can subsequently be `run`."""
+    # Download emulator image.
+    if not environment.get_value('BOT_TMPDIR'):
+      logs.log_error('BOT_TMPDIR is not set.')
+      return
+    temp_directory = environment.get_value('BOT_TMPDIR')
+    archive_src_path = environment.get_value('ANDROID_EMULATOR_BUCKET_PATH')
+    archive_dst_path = os.path.join(temp_directory, 'emulator_bundle.zip')
+    storage.copy_file_from(archive_src_path, archive_dst_path)
+
+    # Extract emulator image.
+    self.emulator_path = os.path.join(temp_directory, 'emulator')
+    archive.unpack(archive_dst_path, self.emulator_path)
+    shell.remove_file(archive_dst_path)
+
+    # Run emulator.
+    script_path = os.path.join(self.emulator_path, 'run')
+    self.process_runner = new_process.ProcessRunner(script_path)
+
+  def run(self):
+    """Actually runs a emulator, assuming `create` has already been called."""
+    if not self.process_runner:
+      raise EmulatorError('Attempted to `run` emulator before calling `create`')
+
+    logs.log('Starting emulator.')
+    self.process = self.process_runner.run(
+        stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+    device_serial = None
+    while not device_serial:
+      line = self.process.popen.stdout.readline().decode()
+      match = DEVICE_SERIAL_RE.match(line)
+      if match:
+        device_serial = match.group(1)
+
+    logs.log('Found serial ID: %s.' % device_serial)
+    environment.set_value('ANDROID_SERIAL', device_serial)
+
+    logs.log('Waiting on device')
+    while adb.get_device_state() != 'device':
```

Comment:
instead of this, i think android.adb.wait_until_fully_booted()

Reply:
That depends on some things that don't exist for emulated devices, specifically the 'sys.boot_completed' property.

Reply:
I also considered using adb.wait_for_device(), but that fails if the device hasn't been registered with ADB yet.